server/builtin/app.go

```diff
@@ -0,0 +1,273 @@
+		DisplayName: AppDisplayName,
+		Description: AppDescription,
+	}
+}
+
+func App(conf config.Config) apps.App {
+	return apps.App{
+		Manifest:    Manifest(conf),
+		BotUserID:   conf.BotUserID,
+		BotUsername: config.BotUsername,
+		GrantedLocations: apps.Locations{
+			apps.LocationCommand,
+		},
+		GrantedPermissions: apps.Permissions{
+			apps.PermissionActAsAdmin,
+		},
+	}
+}
+
+func (a *builtinApp) Roundtrip(_ apps.App, creq apps.CallRequest, async bool) (out io.ReadCloser, err error) {
+	defer func(log utils.Logger) {
+		if x := recover(); x != nil {
+			stack := string(debug.Stack())
+			txt := "Call `" + creq.Path + "` panic-ed."
+			log = log.With(
+				"path", creq.Path,
+				"values", creq.Values,
+				"error", x,
+				"stack", stack,
+			)
+			if creq.RawCommand != "" {
+				txt = "Command `" + creq.RawCommand + "` panic-ed."
+				log.Errorw("Recovered from a panic in a command", "command", creq.RawCommand)
+			} else {
+				log.Errorw("Recovered from a panic in a Call")
+			}
+
+			if a.conf.Get().DeveloperMode {
+				txt += "\n"
+				txt += fmt.Sprintf("Error: **%v**.\n", x)
+				txt += fmt.Sprintf("Stack:\n%v", utils.CodeBlock(stack))
+			} else {
+				txt += "Please check the server logs for more details."
+			}
+			out = nil
+			data, errr := json.Marshal(apps.CallResponse{
+				Type:     apps.CallResponseTypeOK,
+				Markdown: txt,
+			})
+			if errr != nil {
+				err = errr
+				return
+			}
+			err = nil
+			out = ioutil.NopCloser(bytes.NewReader(data))
+		}
+	}(a.conf.Logger())
+
+	f, ok := a.router[creq.Path]
+	if !ok {
+		return nil, utils.NewNotFoundError("%s is not found", creq.Path)
+	}
+	cresp := f(creq)
+	data, err := json.Marshal(cresp)
+	if err != nil {
+		return nil, err
+	}
+	return ioutil.NopCloser(bytes.NewReader(data)), nil
+}
+
+func (a *builtinApp) GetStatic(_ apps.App, path string) (io.ReadCloser, int, error) {
+	return nil, http.StatusNotFound, utils.NewNotFoundError("static support is not implemented")
+}
+
+func mdResponse(format string, args ...interface{}) apps.CallResponse {
+	return apps.CallResponse{
+		Type:     apps.CallResponseTypeOK,
+		Markdown: fmt.Sprintf(format, args...),
+	}
+}
+
+func formResponse(form apps.Form) apps.CallResponse {
+	return apps.CallResponse{
+		Type: apps.CallResponseTypeForm,
+		Form: &form,
+	}
+}
+
+func dataResponse(data interface{}) apps.CallResponse {
+	return apps.CallResponse{
+		Type: apps.CallResponseTypeOK,
+		Data: data,
+	}
+}
+
+func emptyForm(_ apps.CallRequest) apps.CallResponse {
+	return formResponse(apps.Form{})
+}
+
+func submitPath(p string) string {
+	return path.Join(p, "submit")
+}
+
+func formPath(p string) string {
+	return path.Join(p, "form")
+}
+
+func lookupPath(p string) string {
+	return path.Join(p, "lookup")
+}
+
+const SysadminOnly = true
+
+func (a *builtinApp) handle(
+	path string,
+	sysadminOnly bool,
+	submitf func(apps.CallRequest) apps.CallResponse,
+) {
+	a.router[submitPath(path)] = requireSysadmin(sysadminOnly, submitf)
```

Comment:
0/5 seems hard to read that you pass a function "requireSysadmin" to all endpoints, instead of only to the functions that require it. I see that the current code works correctly, but seems confusing at this point.

Reply:
@larkox this resulted in a bigger change than expected, but should be cleaner now.